### PR TITLE
Add New tab page options to settings

### PR DESCRIPTION
**Pull request by simonhong**
With this, user can select which page should be opened in NTP.

Screen Shot 2020-10-23 at 5 41 29 PM
Screen Shot 2020-10-23 at 5 41 40 PM

dark mode blank page:
Screenshot from 2020-10-28 10-47-00

Resolves https://github.com/brave/brave-browser/issues/2999

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`, `npm run gn_check`)
- [x] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Requested a security/privacy review as needed.
- [ ] Added appropriate QA labels (`QA/Yes` or `QA/No`) to the associated issue
- [x] Added appropriate release note labels (`release-notes/include` or `release-notes/exclude`) to the associated issue
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:
`npm run test brave_unit_tests -- --filter=*BraveNewTabTest*`

1. Choose Dashboard...
2. Create NTP by new tab button, shortcut(ex, CMD+T or new tab from app menu)
3. Check brave NTP is opened
4. Choose HomePage
5. If newtab is homepage is on, create NTP is brave NTP
6. If newtab is homepage is off, configured page is opened
7. Choose blankpage
8. Check about:blank is loaded for NTP

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

**Comment by mkarolin** [comment]
This setting seems to also apply to Private and Tor windows. Is this expected?

**Comment by simonhong** [comment]
@mkarolin It's not intentional. @rebron should we only apply this behavior to normal window?

**Comment by bsclifton** [comment]
@simonhong I think we only need to apply this to regular window - Private and Private w/ Tor windows both have the informational page that we would like to show

**Comment by simonhong** [comment]
This option is only effective for normal window.

**Comment by simonhong** [comment]
@rebron @bsclifton , one more question - Should `New Window` also follow this option?
With current PR, only new tab created by new tab command(cmd + T or click new tab button) uses this option.
and new window is launched with one NTP.

**Comment by bsclifton** [comment]
Curious what you think @karenkliu @rebron 
image

To me, it makes more sense to put `New Tab shows` at the top (above the `Customize your Brave Dashboard`)

**Comment by bsclifton** [comment]
> @rebron @bsclifton , one more question - Should `New Window` also follow this option?
> With current PR, only new tab created by new tab command(cmd + T or click new tab button) uses this option.
> and new window is launched with one NTP.

Yes - this new setting should be checked/respected when opening a new window too 😄

**Comment by rebron** [comment]
@karenkliu and I chatted about this yesterday. Placement looks good. 

We should hide the homepage option until a user enables the homepage button. Otherwise, it won't really make sense.

File: chromium_src/chrome/browser/ui/browser_commands.cc
Comment (bsclifton) [inline chromium_src/chrome/browser/ui/browser_commands.cc:20]: Maybe also note as part of this comment:
`// the only change is `new_tab_url` instead of `Gurl()`

**Review by bsclifton — COMMENTED** [review]
Code changes look really good! Notes left about how this should work on new window also

@karenkliu may have some comments too 😄

**Comment by karenkliu** [comment]
@simonhong Hi Simon, after going through the details again, these are some updates to the design:

- Remove "Customize Brave Dashboard" label from the New Tab Page section because with the addition of this dropdown the controls no longer apply to just the dashboard
- The `New tab page shows...` dropdown should go at the top of the New Tab Page section

image

- The widgets go at the bottom of the New Tab Page section, after `Top Sites`
- The options are `Dashboard`, `Homepage`, and `Blank page`
- If the user selects `Blank page`, then the user will get a blank white page in light theme and #17171F in dark theme
- If user selects `Homepage` or `Blank page`, then the `Dashboard` options will not be shown:

image
image

**Comment by simonhong** [comment]
@bsclifton @rebron What should we do in this scenario?
At first, user turned homepage button on and New Tab Page is set as `Homepage`.
Then, user turned off homepage button. How should we change `New Tab page shows` option?

**Comment by mkarolin** [comment]
Agree with @simonhong that it would be more beneficial to have Homepage drop-down option visible not based on the Home Button setting, but rather on whether the homepage URL has been set. However, setting the Home URL is not well-discoverable in the settings right now. 

Why not always have the Homepage drop-down visible and, if selected, show to the user the entry field to set the URL (which would be tied to the same pref as the one under Home Button)?

**Comment by bsclifton** [comment]
I'm OK with always showing `Homepage` given above argument. WDYT @rebron? I could go either way 😛 Would defer to you

**Comment by simonhong** [comment]
@karenkliu When user select `Homepage`, user can choose `New Tab page` as a homepage button option.
Or user can set empty string as a homepage url. In these cases, dashboard options should be visible.

**Comment by simonhong** [comment]
@karenkliu How about below item as a separate issue? I think customizing blank tab color could not be related with this ntp option. IMO, it would be good to use separate bg color for blank page depends on brave theme instead of this ntp option. WDYT?
`If the user selects Blank page, then the user will get a blank white page in light theme and #17171F in dark theme`

File: patches/chrome-browser-ui-browser_commands.cc.patch
Comment (simonhong) [inline patches/chrome-browser-ui-browser_commands.cc.patch]: Difficult to avoid this patching because `OpenEmptyWindow` is used in the same file.

**Comment by simonhong** [comment]
To avoid patching and overriding, I tried url rewriting - https://github.com/brave/brave-core/pull/6879/commits/8dd1bec0a47e2b2e10296372cf41e0efa57d951d
it works fine except blank page option. With blank page option, omnibox and tab title is not set properly like below.
Screen Shot 2020-10-23 at 5 58 07 PM

**Comment by simonhong** [comment]
With navigation throttle (https://github.com/brave/brave-core/pull/6879/commits/9aad04791d2162a7669d273db42dfa6c3e7861bb), all works well.
@mkarolin @bsclifton PTAL again.

**Comment by karenkliu** [comment]
@simonhong 
>How about handling below item as a separate issue? I think customizing blank tab color could not be related with this ntp option. IMO, it would be good to use separate bg color for blank page depends on brave theme regardless of this ntp option.

It's up to you if you want to handle this in a different issue - but we definitely need to have a dark color by default for dark theme. It can't just be white; we're going to get a lot of complaints about it if our blank page is white in dark theme.

**Comment by petemill** [comment]
@simonhong without wanting to change every tab's background color (as likewise we don't want to change every about:blank's background color as it's used in frames, etc), how about the "Blank" option just routes to the normal `BraveNewTabUI : WebUIController` but in that constructor we look at the pref and simply render a custom blank.html instead?

File: browser/new_tab/new_tab_shows_options.cc
Comment (petemill) [inline browser/new_tab/new_tab_shows_options.cc:38]: As commented, how about we use regular existing NTP URL for `BLANKPAGE` option and just have it render a blank page that we can control bg of and anything else we may want to do with a "blank" page?

**Review by petemill — COMMENTED** [review]
Nice use of navigation throttle. Settings all looks good to me. Left a suggestion re: blank page customization.

File: browser/new_tab/new_tab_shows_options.cc
Comment (simonhong) [reply browser/new_tab/new_tab_shows_options.cc:38]: @petemill good idea! 👍

**Comment by bsclifton** [comment]
+1 on nice use of navigation throttle - rewrite was a nice option too (I might use this for a different PR 😄)

**Review by bsclifton — APPROVED** [review]
LGTM (pending feedback from @petemill)

**Comment by rebron** [comment]
> I'm OK with always showing `Homepage` given above argument. WDYT @rebron? I could go either way 😛 Would defer to you

Always show is fine. Thanks for outlining the scenarios @simonhong

**Comment by simonhong** [comment]
@petemill `brave://blank/` webui is introduced. PTAL again.
and ping to @mkarolin for patch/override review :)

**Comment by simonhong** [comment]
CI only have unrelated test failure with this PR - `RewardsContributionBrowserTest.RecurringTipForUnverifiedPublisher`.

File: components/brave_blank_page/resources/brave_blank_page.html
Comment (petemill) [inline components/brave_blank_page/resources/brave_blank_page.html:7]: We can put a `<style>` element here which varies the background-color based on dark / light theme. Similar to on the dashboard NTP: https://github.com/brave/brave-core/blob/master/components/brave_new_tab_ui/brave_new_tab.html

```
@media (prefers-color-scheme: dark) {
    body { background-color: #333639; }
  }
```

Then we don't need _both_ `blank_page.html` and `dark_blank_page.html`.

File: browser/ui/webui/brave_blank_page_ui.h
Comment (petemill) [inline browser/ui/webui/brave_blank_page_ui.h:21]: Instead of creating a whole new WebUI route and class, I was thinking we could just use the regular `chrome://newtab` route and class, and simply change the .html resource that is rendered, depending on the value of `kNewTabPageShowsOptions`. I can't think of a reason that we would need to have separate routes or both the dashboard route and the blank route available at the same time, regardless of the preference value. Do you have a different opinion?

File: browser/ui/webui/brave_blank_page_ui.cc
Comment (petemill) [inline browser/ui/webui/brave_blank_page_ui.cc:18]: See comment about no need to vary on dark mode.

File: browser/new_tab/new_tab_shows_options.cc
Comment (petemill) [inline browser/new_tab/new_tab_shows_options.cc]: see comment about not needing a separate route as far as I can tell

**Review by petemill — CHANGES_REQUESTED** [review]
Can remove the different resources for light and dark. I suggest we just use chrome://newtab and serve blank.html from there, but maybe there's a reason you created chrome://blank

File: browser/ui/webui/brave_blank_page_ui.h
Comment (simonhong) [reply browser/ui/webui/brave_blank_page_ui.h:21]: I thought new tab and blank are different webui. So, I created new route.
But, I don't have strong opinon about this. I'll follow your suggestion.

File: components/brave_blank_page/resources/brave_blank_page.html
Comment (simonhong) [reply components/brave_blank_page/resources/brave_blank_page.html]: Fixed :+1:

File: browser/ui/webui/brave_blank_page_ui.h
Comment (simonhong) [reply browser/ui/webui/brave_blank_page_ui.h]: @petemill  Fixed. PTAL again!

Comment (simonhong) [reply browser/ui/webui/brave_blank_page_ui.h]: I realized that your approach is better because blank page can also act like NTP. - search suggestion in location bar.

**Review by mkarolin — APPROVED** [review]
chromium_src overrides LGTM

File: components/brave_blank_page/resources/brave_blank_new_tab.html
Comment (petemill) [inline components/brave_blank_page/resources/brave_blank_new_tab.html:12]: This end tag doesn't have a start tag

**Review by petemill — DISMISSED** [review]
Looks great to me apart from 1 html syntax error to fix

File: components/brave_blank_page/resources/brave_blank_new_tab.html
Comment (simonhong) [reply components/brave_blank_page/resources/brave_blank_new_tab.html:13]: @petemill Thanks for catching. fixed :)

**Comment by simonhong** [comment]
Merged because only unrelated rewards browser test are failed.

**Pull request merged**
